pkg/controller/awsmachine/awsmachine_controller.go

```diff
@@ -153,116 +154,72 @@ func (r *ReconcileAWSMachine) Reconcile(request reconcile.Request) (reconcile.Re
 }
 
 func (r *ReconcileAWSMachine) reconcile(ctx context.Context, scope *scope.MachineScope) error {
-	exist, err := r.exists(scope)
-	if err != nil {
-		klog.Errorf("Failed to check if Machine %q infrastructure exists: %v", scope.Name(), err)
-		return err
-	}
-
-	// Reconcile ProviderID.
-	if pid := scope.GetProviderID(); pid == nil || *pid == "" {
-		scope.SetProviderID(fmt.Sprintf("aws:////%s", *scope.GetInstanceID()))
-	}
-
-	if exist {
-		scope.Info("Reconciling Machine triggers idempotent update")
-		return r.update(scope)
-	}
-
-	// Machine resource created. Machine does not yet exist.
-	scope.Info("Reconciling Machine triggers idempotent create")
-	return r.create(scope)
-}
-
-// create creates a machine and is invoked by the machine controller.
-func (r *ReconcileAWSMachine) create(scope *scope.MachineScope) error {
 	if scope.Parent.Cluster.Annotations[infrav1.AnnotationClusterInfrastructureReady] != infrav1.ValueReady {
 		scope.Info("Cluster infrastructure is not ready yet - requeuing machine")
 		return &capierrors.RequeueAfterError{RequeueAfter: waitForClusterInfrastructureReadyDuration}
 	}
 
+	var instance *infrav1.Instance
 	ec2svc := ec2.NewService(scope.Parent)
 
-	scope.Info("Retrieving machines for cluster")
-	machineList := &clusterv1.MachineList{}
-	if err := r.List(context.Background(), machineList, scope.Parent.ListOptionsLabelSelector()); err != nil {
-		return errors.Wrapf(err, "failed to retrieve machines in cluster %q", scope.Parent.Name())
-	}
-
-	controlPlaneMachines := util.GetControlPlaneMachinesFromList(machineList)
-	if len(controlPlaneMachines) == 0 {
-		scope.Info("No control plane machines exist yet - requeuing")
-		return &capierrors.RequeueAfterError{RequeueAfter: waitForControlPlaneMachineExistenceDuration}
+	// Parse the ProviderID.
+	pid, err := noderefutil.NewProviderID(scope.GetProviderID())
+	if err == noderefutil.ErrEmptyProviderID {
+		// Query for the instance by tags if the ProviderID is nil.
+		instance, err = ec2svc.InstanceByTags(scope)
+		if err != nil {
+			return errors.Wrapf(err, "failed to query AWSMachine instance by tags")
+		}
+	} else if err != nil {
+		return errors.Wrapf(err, "failed to parse Spec.ProviderID")
 	}
 
-	// Create the Machine.
-	i, err := ec2svc.CreateOrGetMachine(scope)
-	if err != nil {
-		return errors.Errorf("failed to create or get machine: %+v", err)
+	// If the Spec.ProviderID is not set and the instance is not found,
+	// create the AWSMachine.
+	if pid == nil && instance == nil {
```

Comment:
This seems to rely on pid == nil in the case that noderefutil.NewProviderID returns an error, while that is likely always going to be the case it seems to go against Go convention of not relying on any particular value of a return value if an error is returned.

Reply:
I can move this create block in the `err == noderefutil.ErrEmptyProviderID`, although the nesting of if conditions here is definitely a concern for readability